### PR TITLE
Add kernel matched filter

### PR DESCRIPTION
**Pull request by zxdawn**
The kernel matched filter is defined in [Kwon & Nasrabadi (2007)](https:doi.org/10.1007/s11263-006-6689-3).

Briefly,  Given kernel function and target/background matrix, the kernel matched filter response is given by:

```math

\begin{align} 
        y(\hat{k}_r) = \frac{\hat{k}_t^T \hat{K}^{-2} \hat{k}_x}{\hat{k}_t^T \hat{K}^{-2} \hat{k}_t} \\

    where:

       \hat{k}_{t}^T=k_{t}^T-\frac{1}{N} \sum_{i=1}^N k\left(x_i, t\right) \overrightarrow{1} \\
       \hat{k}_{x}^T=k_{x}^T-\frac{1}{N} \sum_{i=1}^N k\left(x_i, x\right) \overrightarrow{1} \\

\end{align}
```

$\hat{K}$ is the centered kernel (Gram) matrix, K = K(X, X) = (K)ij the N × N kernel matrix whose entries are the dot products 〈φ(xi ),φ(x j )〉.

$\hat{k}$ is a kernel, such as linear kernel, polynomial kernel and Gaussian Radial Bases Function kernel (RBF).

The inverse $\hat{K}^{-2}$ may not be numerically stable if the background spectral samples are not independent.
Therefore, the pseudo-inverse of K is used, which is based on eigenvalue decomposition,
where eigenvectors with eigenvalues larger than `eigval_min` are used.

### CODE REVIEW

**Comment by zxdawn** [comment]
I have tested the preliminary version of KMF and found there seems something wrong .... The notebook is on the [gist](https://gist.github.com/zxdawn/a4a980373c5195fdc244fa2ac9978745).

## MF

image

## KMF with poly kernel

Original data:
image

Normalized data:
image

## KMF with rbf kernel

Original data:
image

Normalized data:
image

**Comment by tboggs** [comment]
In the gist notebook, did you intend for your normalization to use an average over rows (as opposed to the entire image)? That is what your function appears to be doing.

**Comment by zxdawn** [comment]
Yes, because most push-broom sensors have
a slightly nonuniform behavior at different cross-track positions. I see someone applied it [here](https://github.com/UCSB-VRL/MethaneMapper-Spectral-Absorption-aware-Hyperspectral-Transformer-for-Methane-Detection/blob/7ae04159e076ca7b5ba4f60c330161bff924b934/train_dataset_generator/train_data_generator.py#L146).

**Comment by zxdawn** [comment]
I have also tested with some small read data and the result is also strange. Here's the [gist](https://gist.github.com/zxdawn/9eec390b75876ffbdc36ef0138267818).
